scvi/models/log_likelihood.py

```diff
@@ -47,31 +47,43 @@ def compute_reconstruction_error(vae, posterior, **kwargs):
 
 def compute_marginal_log_likelihood(vae, posterior, n_samples_mc=100):
     """ Computes a biased estimator for log p(x), which is the marginal log likelihood.
-
-    Despite its bias, the estimator still converges to the real value
-    of log p(x) when n_samples_mc (for Monte Carlo) goes to infinity
-    (a fairly high value like 100 should be enough)
-    Due to the Monte Carlo sampling, this method is not as computationally efficient
-    as computing only the reconstruction loss
+        Despite its bias, the estimator still converges to the real value
+        of log p(x) when n_samples_mc (for Monte Carlo) goes to infinity
+        (a fairly high value like 100 should be enough)
+        Due to the Monte Carlo sampling, this method is not as computationally efficient
+        as computing only the reconstruction loss
     """
     # Uses MC sampling to compute a tighter lower bound on log p(x)
     log_lkl = 0
     for i_batch, tensors in enumerate(posterior):
         sample_batch, local_l_mean, local_l_var, batch_index, labels = tensors
-        x = torch.log(1 + sample_batch)
         to_sum = torch.zeros(sample_batch.size()[0], n_samples_mc)
+
         for i in range(n_samples_mc):
-            qz_m, qz_v, z = vae.z_encoder(x, labels)
-            reconst_loss, kl_divergence = vae(sample_batch, local_l_mean,
-                                              local_l_var,
-                                              batch_index=batch_index,
-                                              y=labels)
+
+            # Distribution parameters and sampled variables
+            px_scale, px_r, px_rate, px_dropout,\
+                qz_m, qz_v, z,\
+                ql_m, ql_v, library = vae.inference(sample_batch, batch_index, labels)
+
+            # Reconstruction Loss
+            if vae.reconstruction_loss == 'zinb':
```

Comment:
Let's not use duplicate code and directly correct the reconstruction error function

Reply:
Something a bit tedious is that on the one hand we are computing the reconstruction loss directly from batch_sample and the parameters of the (ZI)NB in `compute_marginal_log_liklihood` and on the other hand the `compute_reconstruction_error` function takes as an input the VAE and the Posterior object. So we can't use it directly or we would have to make both functions "flow" through the tensors in parallel, which seems to be very complicated.

Can't we just make VAE's `_reconstruction_loss` method public ? Or even create a public wrapper like "get_reconstruction_loss" ?

Reply:
I think it is OK to make _reconstruction_loss public @jeff-regier  any objections ?

Reply:
Ok so I renamed `_reconstruction_loss` to `get_reconstruction_loss` for the coherence with all the "get" and "sample"-like methods in the VAE class.